### PR TITLE
Bug fixes to OSX FSW

### PR DESCRIPTION
**Pull request by sokket**
Fixing bugs in the OSX FileSystemWatcher to make almost all tests pass (minus 3)

@stephentoub @ellismg

### CODE REVIEW

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 1]
```diff
@@ -387,7 +344,7 @@ private bool DoesPathPassNameFilter(string eventPath)
             Interop.EventStream.FSEventStreamEventFlags[] eventFlags,
             FSEventStreamEventId[] eventIds)
         {
-            Debug.Assert((currentIndex + 1 < eventPaths.Length) && (currentIndex + 1 < eventFlags.Length) & (currentIndex + 1 < eventIds.Length));
+            //Debug.Assert((currentIndex + 1 < eventPaths.Length) && (currentIndex + 1 < eventFlags.Length) && (currentIndex + 1 < eventIds.Length));
```
Comment (ellismg) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Is there nothing we can assert here?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 2]
```diff
@@ -295,7 +247,7 @@ private void WatchForFileSystemEventsThreadStart()
                 else if ((DoesPathPassNameFilter(eventPaths[i])) && ((_filterFlags & eventFlags[i]) != 0))
                 {
                     // Remove the base directory prefix (including trailing / that OS X adds)
-                    string relativePath = eventPaths[i].Remove(0, _fullDirectory.Length + 1);
+                    string relativePath = (eventPaths[i].Equals(_fullDirectory, StringComparison.CurrentCultureIgnoreCase) ? _fullDirectory : eventPaths[i].Remove(0, _fullDirectory.Length).TrimStart('/'));
```
Comment (ellismg) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: I think you want to be doing an ordinal comparison here (and I guess ignoring case, if that's what we do other places in OSX land).

File: src/System.IO.FileSystem.Watcher/tests/Utility/Utility.cs [hunk 3]
```diff
@@ -6,6 +6,7 @@
 using System.Threading;
 using System.Runtime.CompilerServices;
 using Xunit;
+using System.Runtime.InteropServices;
```
Comment (ellismg) [inline src/System.IO.FileSystem.Watcher/tests/Utility/Utility.cs]: Seems like you could remove this and leave the file unchanged.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 1 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: The previous assumptions about this assert were incorrect in that we can have a rename without a pair when an item is moved out / in from / to the watcher directory. The current index may be the last item in the array of events, meaning the only thing we could assert is that the index is within the bounds of the array, but we would have other (more fundamental issues) if that ever occurred. I should remove the line though...forgot to do that

**Comment by ellismg** [comment]
I would rebase and squash 9af76dd into the previous commit for a nicer history.

File: src/System.IO.FileSystem.Watcher/tests/Utility/Utility.cs (hunk 3 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/tests/Utility/Utility.cs]: Ah forgot to remove that...thanks, will do

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 4]
```diff
@@ -160,7 +164,7 @@ private void CreateStreamAndStartWatcher()
                     _callback,
                     arrPaths,
                     Interop.EventStream.kFSEventStreamEventIdSinceNow,
-                    0.2f,
+                    0.0f,
```
Comment (ellismg) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Can you help me understand what was going on here before?  It seems like now we just disable coalescing from the OS, was the coalescing causing problems?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 2 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Fixed

File: src/System.IO.FileSystem/src/System/IO/UnixFileSystemObject.cs [hunk 5]
```diff
@@ -178,7 +178,14 @@ public DateTimeOffset CreationTime
...
                         default(DateTimeOffset);
                 }
-                set { } // Not supported
+                set 
+                {
+                    // The ctime in Unix can be interpreted differently by different formats so there isn't
+                    // a reliable way to set this; however, we can't just do nothing since the FileSystemWatcher
+                    // specifically looks for this call to make a Metatdata Change, so we should rev the 
+                    // LastAccessTime of the file to cause the metadata change we need.
+                    LastAccessTime = DateTimeOffset.Now;
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/UnixFileSystemObject.cs]: Is a call to utime sufficient to trigger the event, or do we actually need to change the value?  I'm not loving actually changing the last access time if we can avoid it.  In other words, would this be sufficient to trigger the event?
```C#
LastAccessTime = LastAccessTime;
```
?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Rather than doing a Remove and then a Trim, can we just manually count how many '/'s there are after _fullDirectory.Length and pass that value as the length to Remove?  As it is we're potentially creating two strings when we could instead just do one.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 6]
```diff
@@ -326,23 +280,26 @@ private void WatchForFileSystemEventsThreadStart()
...
                         //       So look for deletes first since those will mean the listening app can't do anything with the item since it doesn't exist (such as stat 
                         //       the file to determine what exactly changed, in this example).
-                        if ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved) == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved)
-                        {
-                            NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, relativePath);
-                        }
-                        else if ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated) == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated)
+                        if ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated) == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated)
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: The length of these fully-qualified consts makes it a bit difficult to read, and the duplication of the name can lead to accidental typos; did you consider adding a helper somewhere like:
```C#
private static bool IsSet(int value, int flags)
{
    return (value & flags) != flags;
}
```
and then using it like:
```C#
if (IsSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated) { ... }
```

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 7]
```diff
@@ -326,23 +280,26 @@ private void WatchForFileSystemEventsThreadStart()
...
+
+                        if (((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemInodeMetaMod)  == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemInodeMetaMod)  ||
+                            ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified)      == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified)      ||
+                            ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemFinderInfoMod) == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemFinderInfoMod) ||
+                            ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner)   == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner)   ||
+                            ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemXattrMod)      == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemXattrMod))
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Such an IsSet could be used here as well.  If all of these flags are a single bit, you could change the IsSet to do ```!= 0``` instead of ```== flags``` and then make all of them in a single call rather than multiple checks and conditionals, e.g.
```C#
if (IsSet(eventFlags[i], 
    kFSEventStreamEventFlagItemInodeMetaMod |
    kFSEventStreamEventFlagItemModified |
    kFSEventStreamEventFlagItemFinderInfoMod |
    ...))
{
    ...
}
```

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 8]
```diff
@@ -304,7 +256,9 @@ private void WatchForFileSystemEventsThreadStart()
                         long pairedId = FindRenameChangePairedChange(i, eventPaths, eventFlags, eventIds);
                         if (pairedId == long.MinValue)
                         {
-                            OnError(new ErrorEventArgs(new ArgumentOutOfRangeException("pairedId")));
+                            // Getting here means we have a rename without a pair, meaning it should be a create for the 
+                            // move from unwatched folder to watcher folder scenario
+                            NotifyFileSystemEventArgs(WatcherChangeTypes.Created, relativePath);
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Could it also be the opposite, that we get a rename for the "from" but not for the "to" because the file moved out of the direction being watched rather than into the directory being watched?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 4 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: This value didn't seem to cause a difference in coalescing...the OS still coalesced events as expected. I'm not 100% sure what this value is supposed to affect since changing it to 0 made no noticeable difference; due to that, making it 0 is much more clear that it isn't some magic value

File: src/System.IO.FileSystem/src/System/IO/UnixFileSystemObject.cs (hunk 5 above)
Comment (sokket) [inline src/System.IO.FileSystem/src/System/IO/UnixFileSystemObject.cs]: Looks like that works...changed to not actually change the time but still trigger the metadata mod

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 7 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Fixed

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 6 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Fixed

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 8 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Fixed the comment

**Comment by sokket** [comment]
@ellismg I'll squash before I put out my final iteration

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 9]
```diff
@@ -295,16 +247,18 @@ private void WatchForFileSystemEventsThreadStart()
...
-                    if ((eventFlags[i] & Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed) == Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed)
+                    if (IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed))
                     {
                         // Find the rename that is paired to this rename, which should be the next rename in the list
                         long pairedId = FindRenameChangePairedChange(i, eventPaths, eventFlags, eventIds);
                         if (pairedId == long.MinValue)
                         {
-                            OnError(new ErrorEventArgs(new ArgumentOutOfRangeException("pairedId")));
+                            // Getting here means we have a rename without a pair, meaning it should be a create for the 
+                            // move from unwatched folder to watcher folder scenario or a move from the watcher folder out
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Your most recent commit updated the comment to include moving out of the watched folder, but it didn't update the implementation accordingly... wouldn't moving out be treated as deletion, not creation?

**Comment by sokket** [comment]
@stephentoub @ellismg updated with squashed version and all tests passing

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 10]
```diff
@@ -292,19 +256,43 @@ private void WatchForFileSystemEventsThreadStart()
...
-                    // Remove the base directory prefix (including trailing / that OS X adds)
-                    string relativePath = eventPaths[i].Remove(0, _fullDirectory.Length + 1);
+                    // The base FileSystemWatcher does a match check against the relative path before combining with 
+                    // the root dir; however, null is special cased to signify the root dir, so check if we should use that.
+                    string relativePath = null;
+                    if (eventPaths[i].Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase) == false)
+                    {
+                        // Check if the event path, with the root directory removed, begins with a / and
+                        // if so, remove it; otherwise, just remove the root path (which contains a trailing /)
+                        if (eventPaths[i][_fullDirectory.Length] == '/')
```
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: @stephentoub Now that I think about it, should this be Path.DirectorySeparatorChar?

**Comment by sokket** [comment]
@dotnet-bot test this please

File: src/Common/src/Interop/OSX/Interop.EventStream.cs [hunk 11]
```diff
@@ -21,7 +21,7 @@ internal static partial class EventStream
         /// <summary>
         /// This constant specifies that we don't want historical file system events, only new ones
         /// </summary>
-        internal const ulong kFSEventStreamEventIdSinceNow = ulong.MaxValue;
+        internal const ulong kFSEventStreamEventIdSinceNow = 0xFFFFFFFFFFFFFFFF;
```
Comment (stephentoub) [inline src/Common/src/Interop/OSX/Interop.EventStream.cs]: Why this change?  Just curious.  I had to count the number of Fs to verify it was correct.

File: src/Common/src/Interop/OSX/libc/Interop.sync.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (stephentoub) [inline src/Common/src/Interop/OSX/libc/Interop.sync.cs]: I believe ```sync``` is POSIX compliant.  Consider putting it in the Unix rather than OSX directory.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 13]
```diff
@@ -155,12 +159,16 @@ private void CreateStreamAndStartWatcher()
                 // Create the callback for the EventStream
                 _callback = new Interop.EventStream.FSEventStreamCallback(FileSystemEventCallback);
 
+                // Make sure the OS file buffer(s) are fully flushed so we don't get events from cached I/O
+                // like we saw in the Unit Tests
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Comments like "like we saw in the Unit Tests" get stale pretty quickly... someone reading this in the not-too-distant-future won't know what that's referring to.  I'd suggest just cutting that phrase; the previous part of the comment does a good job explaining the reasoning.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 14]
```diff
@@ -281,6 +237,14 @@ private void WatchForFileSystemEventsThreadStart()
 
             for (long i = 0; i < numEvents.ToInt32(); i++)
             {
+                // We need to special case the root path for pattern matching, so cache the path this way in case it is the root
+                // and use it for pattern matching. We'll go back to the eventPaths[i] to get the actual path for notifications
+                string path = eventPaths[i];
+                if (path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase))
+                {
+                    path += System.IO.Path.DirectorySeparatorChar.ToString() + ".";
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Consider caching this into a static readonly string, e.g.
```C#
private static readonly string CurrentDirectorySuffix = Path.DirectorySeparatorChar + ".";
...
path += CurrentDirectorySuffix;
```
As it's currently written, not only will string.Concat be called by the compiler to produce a new string, but each call will end up allocating a temporary string containing "/" for the separator.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 10 above)
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Ideally, but it's not a big deal in this case, since the file is already specific to OS X.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 15]
```diff
@@ -455,8 +455,9 @@ private void NotifyFileSystemEventArgs(WatcherChangeTypes changeType, string nam
                     Debug.Fail("Unknown FileSystemEvent change type!  Value: " + changeType);
                     break;
             }
-
-            if (handler != null && MatchPattern(name))
+            
+            bool passedPatternMatch = (string.IsNullOrEmpty(name) ? MatchPattern(_directory) : MatchPattern(name));
+            if (handler != null && passedPatternMatch)
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: This is going to invoke MatchPattern even if handler is null.  Will it be common for handler to be null?  You could instead write it like:
```C#
if (handler != null && MatchPattern(string.IsNullOrEmpty(name) ? _directory : name))
{
    handler(...);
}
```

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs [hunk 16]
```diff
@@ -75,4 +75,22 @@ public static void FileSystemWatcher_Changed_Negative()
             Utility.ExpectNoEvent(eventOccured, "changed");
         }
     }
+
+        [Fact]
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs]: Nit: odd indenting

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs [hunk 17]
```diff
@@ -75,4 +75,22 @@ public static void FileSystemWatcher_Changed_Negative()
...
+            watcher.Filter = "*";
+            AutoResetEvent eventOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Changed);
+
+            watcher.EnableRaisingEvents = true;
+
+            Directory.SetLastAccessTime(watcher.Path, DateTime.Now);
+
+            Utility.ExpectEvent(eventOccured, "changed");
+        }
+    }
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs]: Good, thanks for adding this test.

File: src/Common/src/Interop/OSX/Interop.EventStream.cs (hunk 11 above)
Comment (sokket) [inline src/Common/src/Interop/OSX/Interop.EventStream.cs]: This is just making it exactly what it is in the header file and not having to look up the specific value for MaxValue....it just made it more explicit

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 13 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: Good idea

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 14 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: The second option looks like the best option...fixed

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs (hunk 15 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: Good idea, plus it keeps the if statement mostly like previous versions. Fixed

File: src/Common/src/Interop/OSX/Interop.EventStream.cs (hunk 11 above)
Comment (stephentoub) [inline src/Common/src/Interop/OSX/Interop.EventStream.cs]: Ok.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs (hunk 16 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs]: Oops...

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs (hunk 17 above)
Comment (sokket) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Changed.cs]: :+1:

File: src/Common/src/Interop/OSX/libc/Interop.sync.cs (hunk 12 above)
Comment (sokket) [inline src/Common/src/Interop/OSX/libc/Interop.sync.cs]: So it is :) added to the Unix directory

**Comment by sokket** [comment]
@stephentoub @ellismg latest version passing with all comments fixed

**Comment by ellismg** [comment]
LGTM

**Comment by sokket** [comment]
Thanks all! Merging now
